### PR TITLE
Fixes tracking of lambdas and closures over multiple EnC generations

### PR DESCRIPTION
**Pull request by tmat**
Track the generation in which lambdas and closures were introduced and include it in the synthesized method/type names.

### CODE REVIEW

**Comment by tmat** [comment]
@cston

**Comment by tmat** [comment]
@gafter @VSadov @wschae @cston Could you please review? This is an RC approved change and needs to go in asap.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs [hunk 1]
```diff
@@ -120,30 +120,41 @@ internal static string MakeBaseMethodWrapperName(int uniqueId)
...
+            string suffix = null, 
+            int entityOrdinal = -1, 
+            int entityGeneration = -1,
+            bool isTypeName = false)
         {
             Debug.Assert(methodOrdinal >= -1);
-            Debug.Assert(generation >= 0 || generation == -1 && methodOrdinal == -1);
-            Debug.Assert(uniqueId >= -1);
+            Debug.Assert(methodGeneration >= 0 || methodGeneration == -1 && methodOrdinal == -1);
+            Debug.Assert(entityOrdinal >= -1);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs]: Could this be strengthened to Debug.Assert((methodGeneration >= 0) == (methodOrdinal >= 0))?

Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs]: Could this be strengthened to Debug.Assert((methodGeneration >= 0) == (methodOrdinal >= 0))?

Same comment for entityGeneration and entityOrdinal below.

Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs]: No. Method ordinal can still be -1 for methods that don't have an ordinal (async lambdas).

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs [hunk 2]
```diff
@@ -911,7 +909,33 @@ public override BoundNode VisitConversion(BoundConversion conversion)
...
+        private DebugId GetClosureId(SyntaxNode syntax, ArrayBuilder<ClosureDebugInfo> closureDebugInfo)
+        {
+            Debug.Assert(syntax != null);
+
+            DebugId closureId;
+            DebugId previousClosureId;
+            if (slotAllocatorOpt != null && slotAllocatorOpt.TryGetPreviousClosure(syntax, out previousClosureId))
+            {
+                closureId = previousClosureId;
+            }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: Minor point: Perhaps remove 'previousClosureId' and pass 'closureId' to TryGetPreviousClosure instead. Then the if part can be removed as well. Same comment for VB.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs [hunk 3]
```diff
@@ -940,31 +964,22 @@ private void GetLambdaId(SyntaxNode syntax, ClosureKind closureKind, int closure
...
-            if (slotAllocatorOpt != null &&
-                slotAllocatorOpt.TryGetPreviousLambda(lambdaOrLambdaBodySyntax, isLambdaBody, out previousLambdaOrdinal))
+            DebugId lambdaId;
+            DebugId previousLambdaId;
+            if (slotAllocatorOpt != null && slotAllocatorOpt.TryGetPreviousLambda(lambdaOrLambdaBodySyntax, isLambdaBody, out previousLambdaId))
             {
-                topLevelMethodId = slotAllocatorOpt.PreviousMethodId;
-                lambdaOrdinal = previousLambdaOrdinal;
+                lambdaId = previousLambdaId;
             }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: Perhaps remove 'previousLambdaId' and pass 'lambdaId' instead. Same comment for VB.

**Comment by cston** [comment]
LGTM

**Comment by tmat** [comment]
@dotnet-bot retest this, please.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs [hunk 4]
```diff
@@ -334,17 +331,18 @@ private LambdaFrame GetStaticFrame(DiagnosticBag diagnostics, BoundNode lambda)
 
                 if (_lazyStaticLambdaFrame == null)
                 {
-                    MethodDebugId methodId;
+                    DebugId methodId;
                     if (isNonGeneric)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: I suggest using `?:`

**Pull request merged**